uefi-raw/src/protocol/pci/root_bridge.rs

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use crate::table::boot::{AllocateType, MemoryType};
+use crate::{Handle, PhysicalAddress, Status};
+use core::ffi::c_void;
+use uguid::{guid, Guid};
+
+newtype_enum! {
+    /// Corresponds to the `EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_WIDTH` enum.
+    pub enum PciRootBridgeIoProtocolWidth: u32 => {
+        UINT8 = 0,
+        UINT16 = 1,
+        UINT32 = 2,
+        UINT64 = 3,
+        FIFO_UINT8 = 4,
+        FIFO_UINT16 = 5,
+        FIFO_UINT32 = 6,
+        FIFO_UINT64 = 7,
+        FILL_UINT8 = 8,
+        FILL_UINT16 = 9,
+        FILL_UINT32 = 10,
+        FILL_UINT64 = 11,
+        MAXIMUM = 12,
+    }
+}
+
+newtype_enum! {
+    /// Corresponds to the `EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL_OPERATION` enum.
+    pub enum PciRootBridgeIoProtocolOperation: u32 => {
+        BUS_MASTER_READ = 0,
+        BUS_MASTER_WRITE = 1,
+        BUS_MASTER_COMMON_BUFFER = 2,
+        BUS_MASTER_READ64 = 3,
+        BUS_MASTER_WRITE64 = 4,
+        BUS_MASTER_COMMON_BUFFER64 = 5,
+        MAXIMUM = 6,
+    }
+}
+
+#[repr(C, packed)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct PciIoAddress {
+    pub reg: u8,
+    pub fun: u8,
+    pub dev: u8,
+    pub bus: u8,
+    pub ext_reg: u32,
+}
+
+impl PciIoAddress {
+    #[must_use]
+    pub const fn new(bus: u8, dev: u8, fun: u8) -> Self {
+        Self {
+            bus,
+            dev,
+            fun,
+            reg: 0,
+            ext_reg: 0,
+        }
+    }
+
+    #[must_use]
+    pub const fn with_register(&self, reg: u8) -> Self {
+        let mut addr = *self;
+        addr.reg = reg;
+        addr.ext_reg = 0;
+        addr
+    }
+
+    #[must_use]
+    pub const fn with_extended_register(&self, ext_reg: u32) -> Self {
+        let mut addr = *self;
+        addr.reg = 0;
+        addr.ext_reg = ext_reg;
+        addr
+    }
+}
+
+#[derive(Debug)]
+#[repr(C)]
+pub struct PciRootBridgeIoAccess<TAddr> {
```

Comment:
The uefi-raw crate doesn't normally use generics. I think in the spec the `address` is treated as `u64` in all cases, even though for for pci specifically the `u64` is broken down into subfields, so let's do that and leave higher-level abstractions to the `uefi` crate.

Reply:
In order to get rid of the generics, we could:
```rust
impl Into<u64> for PciIoAddress {
    fn into(self) -> u64 {
        unsafe { core::mem::transmute(self) }
    }
}
```

And then use it like so:
```rust
let addr = PciIoAddress::new(1, 3, 3).with_register(7);
(raw_proto.pci.read)(
    raw_proto,
    PciRootBridgeIoProtocolWidth::UINT32,
    addr.into(), // API takes u64, we convert PciIoAddress to it
    1,
    &mut raw as *mut u32 as *mut c_void
).to_result()?;
```

However, I would like to point out that my use of generics here was the result of a painful lessons-learned from the UEFI spec, due to bad discoverability.
(The first version of this MR was with u64 ... until I actually tried to use it :sweat_smile:.)

Reply:
I don't see the need for `PciIoAddress` in the uefi-raw crate though, unless I'm missing something and that type is defined in the spec? The uefi-raw crate should generally stick close to the spec when creating types, and more 'user friendly' utilities like this can go in `uefi`.

Reply:
Alright, I remved `PciIoAddress` from the MR.